File: src/java/bps-data-generator/src/main/java/com/github/rnowling/bps/datagenerator/cli/Driver.java
package com.github.rnowling.bps.datagenerator.cli;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.OutputStream;
import java.util.Collection;
import java.util.Random;

import com.github.rnowling.bps.datagenerator.datamodels.Pair;
import com.github.rnowling.bps.datagenerator.datamodels.outputs.Transaction;
import com.github.rnowling.bps.datagenerator.datamodels.simulation.Product;


public class Driver
{
	int nStores;
	int nCustomers;
	double simulationTime;
	long seed;
	File outputDir;
	
	private void printUsage()
	{
		String usage = "BigPetStore Data Generator\n" +
				"\n" +
				"Usage: java -jar bps-data-generator-v0.2.java outputDir nStores nCustomers simulationLength [seed]\n" +
				"\n" + 
				"outputDir - (string) directory to write files\n" +
				"nStores - (int) number of stores to generate\n" +
				"nCustomers - (int) number of customers to generate\n" +
				"simulationLength - (float) number of days to simulate\n" +
				"seed - (long) seed for RNG. If not given, one is reandomly generated.\n";
		
		System.out.println(usage);
	}
	
	private void parseArgs(String[] args)
	{
		if(args.length < 4 || args.length > 5)
		{
			printUsage();
			System.exit(1);
		}
		
		outputDir = new File(args[0]);
		if(! outputDir.exists())
		{
			System.err.println("Given path (" + args[0] + ") does not exist.\n");
			printUsage();
			System.exit(1);
		}
		
		if(! outputDir.isDirectory())
		{
			System.err.println("Given path (" + args[0] + ") is not a directory.\n");
			printUsage();
			System.exit(1);
		}
		
		try
		{
			nStores = Integer.parseInt(args[1]);
		}
		catch(Exception e)
		{
			System.err.println("Unable to parse '" + args[1] + "' as an integer for nStores.\n");
			printUsage();
			System.exit(1);
		}
		
		try
		{
			nCustomers = Integer.parseInt(args[2]);
		}
		catch(Exception e)
		{
			System.err.println("Unable to parse '" + args[2] + "' as an integer for nCustomers.\n");
			printUsage();
			System.exit(1);
		}
		
		try
		{
			simulationTime = Double.parseDouble(args[3]);
		}
		catch(Exception e)
		{
			System.err.println("Unable to parse '" + args[3] + "' as a float for simulationLength.\n");
			printUsage();
			System.exit(1);
		}
		
		if(args.length == 5)
		{
			try
			{
				seed = Long.parseLong(args[4]);
			}
			catch(Exception e)
			{
				System.err.println("Unable to parse '" + args[4] + "' as a long for the seed.\n");
				printUsage();
				System.exit(1);
			}
		}
		else
		{
			seed = (new Random()).nextLong();
		}
	}
	
	private void writeTransactions(Collection<Transaction> transactions) throws Exception
	{
		File outputFile = new File(outputDir.toString() + File.separator + "transactions.txt");
		System.out.println(outputFile.toString());
		OutputStream outputStream = new BufferedOutputStream(new FileOutputStream(outputFile));
		
		for(Transaction transaction : transactions)
		{
			for(Product product : transaction.getProducts())
			{
				String record = transaction.getId() + ",";
				record += transaction.getDateTime() + ",";
				record += transaction.getStore().getId() + ",";
				record += transaction.getStore().getLocation().getZipcode() + ",";
				record += transaction.getCustomer().getId() + ",";
				Pair<String, String> name = transaction.getCustomer().getName();
				record += name.getFirst() + " " + name.getSecond() + ",";
				record += transaction.getCustomer().getLocation().getZipcode() + ",";
				record += product.toString() + "\n";
				
				outputStream.write(record.getBytes());
			}
		}
		
		outputStream.close();
	}
	
	private void run() throws Exception
	{
		Simulation simulation = new Simulation(nStores, nCustomers, simulationTime, seed);
		
		simulation.simulate();
		
		writeTransactions(simulation.getTransactions());
	}	
	public void run(String[] args) throws Exception
	{
		parseArgs(args);
		
		run();
	}
	
	public static void main(String[] args) throws Exception
	{
		Driver driver = new Driver();
		driver.run(args);		
	}
}

File: spark_driver/src/main/scala/com/github/rnowling/bps/datagenerator/spark/Driver.scala
package com.github.rnowling.bps.datagenerator.spark

import com.github.rnowling.bps.datagenerator.datamodels.{Store,Customer,PurchasingProfile,Transaction}
import com.github.rnowling.bps.datagenerator.{DataLoader,StoreGenerator,CustomerGenerator,PurchasingProfileGenerator,TransactionGenerator}
import com.github.rnowling.bps.datagenerator.framework.SeedFactory
import org.apache.spark.{SparkContext, SparkConf}
import org.apache.spark.SparkContext._
import java.util.ArrayList
import scala.util.Random
import java.io.File

object SparkDriver {
  private var nStores: Int = -1
  private var nCustomers: Int = -1
  private var simulationLength: Double = -1.0
  private var seed: Long = -1
  private var outputDir: File = new File(".")

  private val NPARAMS = 5

  private def printUsage() {
    val usage: String = "BigPetStore Data Generator\n" +
      "\n" +
      "Usage: spark-submit ... outputDir nStores nCustomers simulationLength [seed]\n" +
      "\n" + 
      "outputDir - (string) directory to write files\n" +
      "nStores - (int) number of stores to generate\n" +
      "nCustomers - (int) number of customers to generate\n" +
      "simulationLength - (float) number of days to simulate\n" +
      "seed - (long) seed for RNG. If not given, one is reandomly generated.\n"

    println(usage)
  }

  private def parseArgs(args: Array[String]) {
    if(args.length != NPARAMS && args.length != (NPARAMS - 1)) {
      printUsage()
      System.exit(1)
    }

    var i = 0

    outputDir = new File(args(i))
    if(! outputDir.exists()) {
      System.err.println("Given path (" + args(i) + ") does not exist.\n")
      printUsage()
      System.exit(1)
    }

    if(! outputDir.isDirectory()) {
      System.err.println("Given path (" + args(i) + ") is not a directory.\n")
      printUsage()
      System.exit(1)
    }

    i += 1
    try {
      nStores = args(i).toInt
    }
    catch {
      case _ : NumberFormatException =>
        System.err.println("Unable to parse '" + args(i) + "' as an integer for nStores.\n")
        printUsage()
        System.exit(1)
    }

    i += 1
    try {
      nCustomers = args(i).toInt
    }
    catch {
      case _ : NumberFormatException =>
        System.err.println("Unable to parse '" + args(i) + "' as an integer for nCustomers.\n")
        printUsage()
        System.exit(1)
    }

    i += 1
    try {
      simulationLength = args(i).toDouble
    }
    catch {
      case _ : NumberFormatException =>
        System.err.println("Unable to parse '" + args(i) + "' as a float for simulationLength.\n")
        printUsage()
        System.exit(1)
    }

    if(args.length == NPARAMS) {
      i += 1
      try {
        seed = args(i).toLong
      }
      catch {
        case _ : NumberFormatException =>
          System.err.println("Unable to parse '" + args(i) + "' as a long for seed.\n")
          printUsage()
          System.exit(1)
      }
    }
    else {
      seed = (new Random()).nextLong
    }
  }


  def main(args: Array[String]) {
    parseArgs(args)

    val inputData = new DataLoader().loadData()
    val seedFactory = new SeedFactory(seed);

    println("Generating stores...")
    val stores : ArrayList[Store] = new ArrayList()
    val storeGenerator = new StoreGenerator(inputData, seedFactory);
    for(i <- 1 to nStores) {
      val store = storeGenerator.generate()
      stores.add(store)
    }
    println("Done.")

    println("Generating customers...")
    var customers: List[Customer] = List()
    val custGen = new CustomerGenerator(inputData, stores, seedFactory)
    for(i <- 1 to nCustomers) {
      val customer = custGen.generate()
      customers = customer :: customers
    }
    println("Done.")

    val conf = new SparkConf()
      .setAppName("BPS Data Generator")

    val sc = new SparkContext(conf)

    val storesBC = sc.broadcast(stores)
    val productBC = sc.broadcast(inputData.getProductCategories())
    val customerRDD = sc.parallelize(customers)
    val nextSeed = seedFactory.getNextSeed()

    val transactionRDD = customerRDD.mapPartitionsWithIndex { (index, custIter) =>
      val seedFactory = new SeedFactory(nextSeed ^ index)
      val transactionIter = custIter.map{ customer =>
	val products = productBC.value

        val profileGen = new PurchasingProfileGenerator(products, seedFactory)
        val profile = profileGen.generate()

        val transGen = new TransactionGenerator(customer, profile, storesBC.value, products,
          seedFactory)

        var transactions : List[Transaction] = List()
	var transaction = transGen.generate()
        while(transaction.getDateTime() < simulationLength) {
          transactions = transaction :: transactions

          transaction = transGen.generate()
        }

	transactions
      }
      transactionIter
    }.flatMap( s => s)

    println("Generating transactions...")
    val nTrans = transactionRDD.count()
    println(s"Generated $nTrans transactions.")

    val transactionStringsRDD = transactionRDD.map { t =>
      var records : List[String] = List()
      val products = t.getProducts()
      for(i <- 0 until products.size()) {
        val p = products.get(i)
        var record = t.getId() + ","
        record += t.getDateTime() + ","
        record += t.getStore().getId() + ","
        record += t.getStore().getLocation().getZipcode() + ","
        record += t.getCustomer().getId() + ","
	val name = t.getCustomer().getName()
	record += name.getFirst() + " " + name.getSecond() + ","
	record += t.getCustomer().getLocation().getZipcode() + ","
	record += p

        records = record :: records
      }

      records
    }.flatMap { s => s }

    transactionStringsRDD.saveAsTextFile(outputDir + "/transactions.txt")

    sc.stop()
  }
}
